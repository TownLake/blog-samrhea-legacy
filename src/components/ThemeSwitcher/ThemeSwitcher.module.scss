@use "sass:color";
@import "src/assets/scss/mixins";
@import "src/assets/scss/variables";

.themeSwitcher {
  --color: var(--color-theme-switcher);

  .button {
    align-items: center;
    background: transparent;
    border: 0;
    cursor: pointer;
    display: flex;
    justify-content: center;
    outline: 0;

    svg {
      pointer-events: none;
      stroke: var(--color);
      transition: stroke 400ms;
    }

    &:hover {
      --color: var(--color-theme-switcher-hover);
    }
  }

  .moon {
    opacity: 0;
    stroke-dasharray: 0 1px;
    transition: stroke-dasharray 200ms ease-in, opacity 400ms ease-in;
  }

  .sun {
    opacity: 1;
    stroke-dasharray: 1px 1px;
    transition: stroke-dasharray 200ms ease-in, opacity 400ms ease-in;
  }

  &.dark {
    .moon {
      opacity: 1;
      stroke-dasharray: 1px 1px;
    }

    .sun {
      opacity: 0;
      stroke-dasharray: 0 1px;
    }
  }
}
