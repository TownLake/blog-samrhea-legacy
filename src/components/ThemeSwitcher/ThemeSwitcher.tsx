import React, { useEffect, useState } from "react";

import cn from "classnames";

import { useTheme } from "@/hooks";

import * as styles from "./ThemeSwitcher.module.scss";

const ThemeSwitcher: React.FC = () => {
  const [{ mode }, toggleTheme] = useTheme();
  const [isClient, setIsClient] = useState<boolean>(false);

  useEffect(() => {
    setIsClient(typeof window !== "undefined");
  }, []);

  if (!isClient) {
    return null;
  }

  return (
    <div
      className={cn(styles.themeSwitcher, {
        [styles.dark]: mode === "dark",
      })}
    >
      <button className={styles.button} onClick={toggleTheme}>
        <svg
          width="24"
          height="24"
          viewBox="0 0 24 24"
          fill="none"
          strokeWidth="2"
          strokeLinecap="round"
          strokeLinejoin="round"
        >
          <title>{mode}</title>
          <path
            pathLength="1"
            className={styles.moon}
            d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"
          />
          <circle pathLength="1" className={styles.sun} cx="12" cy="12" r="5" />
          <line
            pathLength="1"
            className={styles.sun}
            x1="12"
            y1="1"
            x2="12"
            y2="3"
          />
          <line
            pathLength="1"
            className={styles.sun}
            x1="12"
            y1="21"
            x2="12"
            y2="23"
          ></line>
          <line
            pathLength="1"
            className={styles.sun}
            x1="4.22"
            y1="4.22"
            x2="5.64"
            y2="5.64"
          />
          <line
            pathLength="1"
            className={styles.sun}
            x1="18.36"
            y1="18.36"
            x2="19.78"
            y2="19.78"
          />
          <line
            pathLength="1"
            className={styles.sun}
            x1="1"
            y1="12"
            x2="3"
            y2="12"
          />
          <line
            pathLength="1"
            className={styles.sun}
            x1="21"
            y1="12"
            x2="23"
            y2="12"
          />
          <line
            pathLength="1"
            className={styles.sun}
            x1="4.22"
            y1="19.78"
            x2="5.64"
            y2="18.36"
          />
          <line
            pathLength="1"
            className={styles.sun}
            x1="18.36"
            y1="5.64"
            x2="19.78"
            y2="4.22"
          />
        </svg>
      </button>
    </div>
  );
};

export default ThemeSwitcher;
