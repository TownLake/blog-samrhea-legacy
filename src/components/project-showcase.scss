.project-showcase {
    max-width: 1200px;
    margin: 0 auto;
    padding: 2rem;
  
    h1 {
      text-align: center;
      margin-bottom: 2rem;
      font-size: 2rem;
      font-weight: bold;
    }
  
    .project-grid {
      display: grid;
      grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
      gap: 2rem;
    }
  }
  
  // Dark mode styles
  @media (prefers-color-scheme: dark) {
    .project-showcase {
      background-color: #1a202c;
      color: #f7fafc;
    }
  }