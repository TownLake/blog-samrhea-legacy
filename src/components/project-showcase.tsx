.project-tile {
    position: relative;
    width: 100%;
    height: 0;
    padding-bottom: 75%; // 4:3 aspect ratio
    overflow: hidden;
    border-radius: 8px;
    box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
    transition: transform 0.3s ease;
  
    &:hover {
      transform: translateY(-5px);
    }
  
    .project-link {
      text-decoration: none;
      color: inherit;
    }
  
    .project-image {
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
      background-size: cover;
      background-position: center;
    }
  
    .project-overlay {
      position: absolute;
      bottom: 0;
      left: 0;
      right: 0;
      background: rgba(0, 0, 0, 0.7);
      color: white;
      padding: 1rem;
      transform: translateY(100%);
      transition: transform 0.3s ease;
    }
  
    &:hover .project-overlay {
      transform: translateY(0);
    }
  
    .project-title {
      margin: 0 0 0.5rem;
      font-size: 1.2rem;
    }
  
    .project-description {
      margin: 0;
      font-size: 0.9rem;
      opacity: 0.8;
    }
  }