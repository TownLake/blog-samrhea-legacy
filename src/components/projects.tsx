import React from 'react';
import { Link } from 'gatsby';
import './project.scss';

interface ProjectProps {
  title: string;
  description: string;
  imageUrl: string;
  projectUrl: string;
}

const Project: React.FC<ProjectProps> = ({ title, description, imageUrl, projectUrl }) => {
  return (
    <div className="project-tile">
      <Link to={projectUrl} className="project-link">
        <div className="project-image" style={{ backgroundImage: `url(${imageUrl})` }}>
          <div className="project-overlay">
            <h3 className="project-title">{title}</h3>
            <p className="project-description">{description}</p>
          </div>
        </div>
      </Link>
    </div>
  );
};

export default Project;