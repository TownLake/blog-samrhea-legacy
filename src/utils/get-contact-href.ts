import { btoa } from "abab";

const getContactHref = (name: string, contact: string) => {
  const hrefs: { [key: string]: string } = {
    email: btoa(contact) || "",
    line: `line://ti/p/${contact}`,
    telegram: `https://t.me/${contact}`,
    vkontakte: `https://vk.com/${contact}`,
    medium: `https://medium.com/${contact}`,
    github: `https://github.com/${contact}`,
    weibo: `https://www.weibo.com/${contact}`,
    gitlab: `https://www.gitlab.com/${contact}`,
    codepen: `https://www.codepen.io/${contact}`,
    twitter: `https://www.twitter.com/${contact}`,
    soundcloud: `https://soundcloud.com/${contact}`,
    facebook: `https://www.facebook.com/${contact}`,
    instagram: `https://www.instagram.com/${contact}`,
    linkedin: `https://www.linkedin.com/in/${contact}`,
    youtube: `https://www.youtube.com/channel/${contact}`,
    mastodon: `${contact}`,
  };

  return hrefs[name] ?? contact;
};

export default getContactHref;
