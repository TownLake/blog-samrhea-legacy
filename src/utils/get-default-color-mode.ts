const getDefaultColorMode = (): "dark" | "light" => {
  if (typeof window === "undefined") {
    return "light";
  }

  const mql = window.matchMedia("(prefers-color-scheme: dark)");
  const hasMediaQueryPreference = typeof mql.matches !== "undefined";
  if (hasMediaQueryPreference) {
    return mql.matches ? "dark" : "light";
  }

  return "light";
};

export default getDefaultColorMode;
