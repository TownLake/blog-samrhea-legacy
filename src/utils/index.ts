export { default as getIcon } from "./get-icon";
export { default as getContactHref } from "./get-contact-href";
export { default as toKebabCase } from "./to-kebab-case";
export { default as getDefaultColorMode } from "./get-default-color-mode";
export * as testUtils from "./test-utils";
