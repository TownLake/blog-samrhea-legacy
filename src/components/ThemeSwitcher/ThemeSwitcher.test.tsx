import React from "react";

import { fireEvent } from "@testing-library/dom";

import { ThemeSwitcher } from "@/components/ThemeSwitcher";
import { testUtils } from "@/utils";

describe("ThemeSwitcher", () => {
  test("renders correctly", () => {
    const tree = testUtils.createSnapshotsRenderer(<ThemeSwitcher />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("dark theme is set correctly", () => {
    window.localStorage.setItem(
      "diesel:theme-atom",
      JSON.stringify({ mode: "dark" }),
    );

    const { getByTitle } = testUtils.renderWithCoilProvider(<ThemeSwitcher />);

    window.localStorage.removeItem("diesel:theme-atom");

    expect(getByTitle("dark")).not.toBeNull();
  });

  test("light theme is set correctly", () => {
    const { getByTitle } = testUtils.renderWithCoilProvider(<ThemeSwitcher />);

    expect(getByTitle("light")).not.toBeNull();
  });

  test("theme switching works correctly", () => {
    const { getByTitle, getByRole } = testUtils.renderWithCoilProvider(
      <ThemeSwitcher />,
    );

    expect(getByTitle("light")).not.toBeNull();

    const button = getByRole("button");

    fireEvent.click(button);

    expect(getByTitle("dark")).not.toBeNull();
  });
});
