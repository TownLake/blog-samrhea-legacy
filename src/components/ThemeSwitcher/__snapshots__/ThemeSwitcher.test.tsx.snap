// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSwitcher renders correctly 1`] = `null`;
